### DISCUSSION TITLE
React folder architechture

### DISCUSSION THREAD

**Discussion by elsycharles** [community/community#133899] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Question

### Body

Did someone know a perfect folder and file architechture for react project to have a clean code ?

**Comment by akibhaider** [comment]
YourApplicationName/
│
├── public/
│   ├── index.html
│   └── assets/          # Static assets (images, fonts, etc.)
│
├── src/
│   ├── assets/          
│   ├── components/      # Reusable components
│   │   ├── ComponentName/
│   │   │   ├── ComponentName.jsx
│   │   │   ├── ComponentName.css (or .scss)
│   │   │   └── index.js
│   ├── hooks/           # Custom hooks
│   │   ├── useHookName.js
│   ├── pages/           # Page components (correspond to routes)
│   │   ├── PageName/
│   │   │   ├── PageName.jsx
│   │   │   ├── PageName.css (or .scss)
│   │   │   └── index.js
│   ├── services/        # API calls or services
│   │   ├── apiService.js
│   ├── store/           # State management (Redux, Context API)
│   │   ├── index.js
│   ├── styles/          # Global styles
│   │   ├── variables.scss (or variables.css)
│   │   ├── global.scss (or global.css)
│   ├── utils/           # Utility functions
│   │   ├── helpers.js
│   ├── App.jsx          # Main app component
│   ├── index.js         # Entry point
│   ├── routes.js        # Routing setup
│   └── config/          # Configuration files
│       ├── config.js
│
├── .env                 # Environment variables
├── .gitignore
├── package.json
└── README.md

Consider following this. I used this in my latest typescript react project

**Reply by elsycharles** [reply]
Thanks for your answer, I will try this in my following project.
but I have a question, do you combine the props interface in the component or create a new file for this ? and if you create a new file for props where you will place it ?

**Reply by akibhaider** [reply]
Depends on your project's demand actually. For smaller project combining the props interface and keeping it inline is okay. For larger projects, you can create your files under a centralized folder. Best practice to me is create a new "types" directory and put those all in it.  Hope it helps

**Reply by wlb-tycoon** [reply]
> Thanks for your answer, I will try this in my following project. but I have a question, do you combine the props interface in the component or create a new file for this ? and if you create a new file for props where you will place it ?

You can define the props interface directly within the component file. This approach is often simpler and more straightforward for smaller components. Here’s an example:
```
// MyComponent.tsx

import React from 'react';

interface MyComponentProps {
  title: string;
  onClick: () => void;
}

const MyComponent: React.FC<MyComponentProps> = ({ title, onClick }) => {
  return (
    <div>
      <h1>{title}</h1>
      <button onClick={onClick}>Click Me</button>
    </div>
  );
};

export default MyComponent;
```
For larger components or when the props interface is reused across multiple components, creating a separate file for the interface can help maintain organization and improve readability. Here’s how you can do it:
Create a Separate File: You can create a **types** or **interfaces** folder in your project structure. Inside this folder, you can create a file like **MyComponent.types.ts** or **MyComponentProps.ts**.
```
src/
├── components/
│   ├── MyComponent.tsx
│   └── ...
└── types/
    └── MyComponent.types.ts
```
Define the Interface in the Separate File:
```
// types/MyComponent.types.ts

export interface MyComponentProps {
  title: string;
  onClick: () => void;
}
```
Import the Interface in Your Component:
```
// MyComponent.tsx

import React from 'react';
import { MyComponentProps } from '../types/MyComponent.types';

const MyComponent: React.FC<MyComponentProps> = ({ title, onClick }) => {
  return (
    <div>
      <h1>{title}</h1>
      <button onClick={onClick}>Click Me</button>
    </div>
  );
};

export default MyComponent;
```
To sum up:
If your component is small and the props interface is not reused, it’s perfectly fine to define the interface within the component file.
For larger components or when the interface might be reused, it’s better to define it in a separate file for better organization.

**Comment by Sivamani-18** [comment]
A well-structured folder and file architecture in a React project can significantly enhance code readability, maintainability, and scalability. Here's a commonly recommended structure that you can adapt based on the specific needs of your project:

### 1. Root Directory

```
project-root/
│
├── public/
│   └── index.html
│
├── src/
│   ├── assets/
│   │   ├── images/
│   │   └── styles/
│   ├── components/
│   │   ├── common/
│   │   │   └── Button.jsx
│   │   ├── layout/
│   │   │   └── Header.jsx
│   │   └── specific/
│   │       └── UserCard.jsx
│   ├── contexts/
│   │   └── AuthContext.jsx
│   ├── hooks/
│   │   └── useAuth.js
│   ├── pages/
│   │   ├── Home.jsx
│   │   └── About.jsx
│   ├── services/
│   │   └── api.js
│   ├── utils/
│   │   └── helpers.js
│   ├── App.jsx
│   ├── index.js
│   └── routes/
│       └── AppRoutes.jsx
│
├── .env
├── .gitignore
├── package.json
├── README.md
└── webpack.config.js
```

### 2. Detailed Explanation

- **public/**: Contains static files like the main HTML file and other assets that won't change during runtime.
  
- **src/**: The main directory for all source code.
  
  - **assets/**: Contains static assets like images and stylesheets.
  
    - **images/**: Stores image files.
    - **styles/**: Stores global CSS/SCSS files.
  
  - **components/**: Contains all reusable React components.
    
    - **common/**: Generic components used throughout the application (e.g., buttons, form inputs).
    - **layout/**: Components related to the layout (e.g., Header, Footer).
    - **specific/**: Components that are specific to certain features or pages (e.g., UserCard, ProductList).
  
  - **contexts/**: Contains context providers and context-related hooks (e.g., AuthContext).
  
  - **hooks/**: Custom hooks for reuse across components (e.g., useAuth).
  
  - **pages/**: Contains page components that represent different routes (e.g., Home, About).
  
  - **services/**: Handles API calls and any external services interaction (e.g., api.js).
  
  - **utils/**: Utility functions and helpers (e.g., formatDate, calculateAge).
  
  - **routes/**: Contains route definitions and configurations (e.g., AppRoutes).

  - **App.jsx**: The main App component that sets up the structure of the application.
  
  - **index.js**: The entry point of the React application.
  
- **.env**: Environment variables.
  
- **.gitignore**: Specifies files and directories to be ignored by Git.
  
- **package.json**: Manages project dependencies and scripts.
  
- **README.md**: Project documentation.
  
- **webpack.config.js**: Configuration file for Webpack (if you're using Webpack).

### 3. Tips for Clean Code

- **Component Naming**: Use clear and descriptive names for components.
- **Component Size**: Keep components small and focused on a single responsibility.
- **Folder Naming**: Use consistent and meaningful folder names.
- **CSS-in-JS**: Consider using CSS-in-JS libraries like styled-components for scoped styling.
- **Testing**: Include a `tests` folder or place test files alongside corresponding components (e.g., `Header.test.jsx`).
- **TypeScript**: Consider using TypeScript for better type safety.

Adopting and consistently following a structure like this will help keep your React codebase clean, organized, and easy to navigate.

**Comment by Mjid1** [comment]
my-react-app/
├── public/
│   ├── index.html
│   └── ...
├── src/
│   ├── assets/
│   │   ├── images/
│   │   │   ├── logo.png
│   │   │   └── background.jpg
│   │   └── fonts/
│   │       └── OpenSans-Regular.ttf
│   ├── components/
│   │   ├── Button/
│   │   │   ├── Button.js
│   │   │   ├── Button.css
│   │   │   └── Button.test.js
│   │   └── Header/
│   │       ├── Header.js
│   │       └── Header.css
│   ├── hooks/
│   │   ├── useFetch.js
│   │   └── useAuth.js
│   ├── pages/
│   │   ├── HomePage.js
│   │   └── AboutPage.js
│   ├── services/
│   │   ├── api.js
│   │   └── authService.js
│   ├── store/
│   │   ├── actions.js
│   │   ├── reducers.js
│   │   └── store.js
│   ├── styles/
│   │   ├── global.css
│   │   └── theme.css
│   ├── utils/
│   │   ├── formatDate.js
│   │   └── calculateSum.js
│   ├── App.js
│   ├── index.js
│   └── ...
├── .gitignore
├── package.json
├── README.md
└── ...

**Comment by mayurmewada** [comment]
appName/
│
├── public/
│   ├── index.html
│   └── assets/
│       ├── fonts/
│       ├── images/
│       └── styles/
│
├── src/
│   ├── components/
│   │   ├── common/
│   │   │   ├── Button.js
│   │   │   └── Input.js
│   │   └── shared/
│   │       ├── Navbar.js
│   │       └── Footer.js
│   │
│   ├── pages/
│   │   ├── Login.js
│   │   ├── Home.js
│   │   └── layout.js
│   │
│   ├── redux/
│   │   ├── store.js
│   │   └── slices/
│   │       ├── authSlice.js
│   │       └── userSlice.js
│   │
│   ├── extra folder/ 
│   │   ├── extra file
│   │   └── extra file
│   │
│   ├── AppRoutes.jsx
│   └── index.js
│
├── .env
├── .gitignore
├── package.json
└── README.md

Simple structure used by most developers

**Comment by UvejsMikullovci** [comment]
Honestly i like the atomic structure so for example if you have a page that has a section with 3 cards that each have smaller components each side, id create a components folder then inside add a atoms, molecules, organisms and id put for example the img h1 and desc of the cards in atoms, the card itself in molecules then the section with the cards in organisms then the section in the organisms folder would be sent to the page this would make it easy for large scale apps you can use promps to the molecules to change the text each time you use it to make them reusable 

<img width="182" alt="screenshot" src="https://github.com/user-attachments/assets/c013bb6b-07d6-4197-9e09-434eec044232" />

**Reply by UvejsMikullovci** [reply]
BTW i am a begginer to coding but this is how i usualy do it soooooooo yeah
